stringtemplate4/src/main/java/org/jdbi/v3/stringtemplate4/StringTemplateSqlLocator.java

```diff
@@ -0,0 +1,117 @@
+import java.io.Reader;
+import java.nio.charset.Charset;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import net.jodah.expiringmap.ExpirationPolicy;
+import net.jodah.expiringmap.ExpiringMap;
+import org.stringtemplate.v4.ST;
+import org.stringtemplate.v4.STGroup;
+import org.stringtemplate.v4.STGroupString;
+
+/**
+ * Locates SQL in <code>.sql.stg</code> StringTemplate group files on the classpath.
+ */
+public class StringTemplateSqlLocator {
+    private static final Map<Class<?>, STGroup> CACHE = ExpiringMap.builder()
+            .expiration(10, TimeUnit.MINUTES)
+            .expirationPolicy(ExpirationPolicy.ACCESSED)
+            .entryLoader(o -> readStringTemplateGroup((Class) o))
+            .build();
+
+    private static final Charset UTF_8 = Charset.forName("UTF-8");
+    private static final String TEMPLATE_GROUP_EXTENSION = ".sql.stg";
+
+    private StringTemplateSqlLocator() {
+    }
+
+    /**
+     * Locates SQL for the given type and name. Example: Given a type <code>com.foo.Bar</code> and a name of
+     * <code>baz</code>, loads a StringTemplate group file from the resource named <code>com/foo/Bar.sql.stg</code> on
+     * the classpath, and returns the template with the given name from the group.
+     *
+     * @param type the type that "owns" the given StringTemplate group file. Dictates the filename of the
+     *             StringTemplate group file on the classpath.
+     * @param name the template name in the StringTemplate group.
+     * @return the located SQL.
+     */
+    public static ST findStringTemplate(Class<?> type, String name) {
+        STGroup group = findStringTemplateGroup(type);
+
+        if (!group.isDefined(name)) {
+            throw new IllegalStateException("No StringTemplate group " + name + " for class " + type);
+        }
+
+        return group.getInstanceOf(name);
+    }
+
+    /**
+     * Loads the StringTemplateGroup for the given type. Example: Given a type <code>com.foo.Bar</code>, returns a
+     * StringTemplateGroup loaded from the resource named <code>com/foo/Bar.sql.stg</code> on the classpath.
+     *
+     * @param type the type that "owns" the given StringTemplate group file. Dictates the filename of the
+     *             StringTemplate group file on the classpath.
+     * @return the loaded StringTemplateGroup.
+     */
+    public static STGroup findStringTemplateGroup(Class<?> type) {
+        return CACHE.get(type);
+    }
+
+    private static STGroup readStringTemplateGroup(Class<?> type) {
+        String path = resourcePathFor(type);
+
+        try (InputStream is = openStream(type.getClassLoader(), path)) {
+            return new STGroupString(toString(is));
+        }
+        catch (IOException e) {
+            throw new RuntimeException("Unable to read StringTemplate group file at " + path + " on classpath", e);
+        }
+    }
+
+    private static String toString(InputStream inputStream) throws IOException {
```

Comment:
I could use Guava, but we've had a longstanding goal to eliminate that dependency if possible.

Maybe it won't be a big deal once we get shaded jars working again.